ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs

```diff
@@ -0,0 +1,239 @@
+    -- index files are truncated to the last valid state, the state before the
+    -- write.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+    --
+    -- TODO the given binary blob may not be empty.
+  , appendBinaryBlob
+      :: HasCallStack
+      => RelativeSlot -> Builder -> m (Either ImmutableDBError ())
+
+    -- | Close the current epoch and start a new epoch (the epoch number is
+    -- incremented by 1). The new epoch number is returned.
+    --
+    -- The size of the new epoch must be passed.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+  , startNewEpoch
+      :: HasCallStack => EpochSize -> m (Either ImmutableDBError Epoch)
+      -- TODO For additional safety in concurrent usage: let the user pass the
+      -- new 'Epoch' to start, which must be 'getCurrentEpoch' + 1, so no
+      -- skipping allowed.
+
+    -- | When given a start (first argument) and a stop (second argument)
+    -- position (both inclusive bounds), return an 'Iterator' to efficiently
+    -- stream binary blocks out of the database.
+    --
+    -- Use 'iteratorNext' to stream from the iterator.
+    --
+    -- Relative slots (and even whole epochs) that do not store a blob are
+    -- skipped by the iterator.
+    --
+    -- Throws an 'InvalidIteratorRangeError' if the start of the range is
+    -- greater than the end of the range.
+    --
+    -- Throws a 'ReadFutureSlotError' if the start or end 'EpochSlot' are in
+    -- the future.
+    --
+    -- Throws a 'SlotGreaterThanEpochSizeError' if the relative slot of the
+    -- start or end 'EpochSlot' exceeds the corresponding epoch size.
+    --
+    -- The iterator is automatically closed when exhausted, and can be
+    -- prematurely closed with 'iteratorClose'.
+  , streamBinaryBlobs
+      :: HasCallStack
+      => EpochSlot
+      -> EpochSlot
+      -> m (Either ImmutableDBError (Iterator m))
+  }
+
+-- | An 'Iterator' is a handle which can be used to efficiently stream binary
+-- blobs across multiple epochs. Slots not containing a blob are skipped.
+data Iterator m = Iterator
+  { -- | Steps an 'Iterator', yielding either an 'ImmutableDBError' or an
+    -- 'IteratorResult'.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+    --
+    -- TODO NOTE(adn): This works under the assumption that the user is the
+    -- exclusive owner of the iterator.
+    --
+    -- When the iterator is exhausted ('IteratorExhausted'), the iterator is
+    -- automatically closed with 'iteratorClose'.
+    iteratorNext :: HasCallStack => m (Either ImmutableDBError IteratorResult)
+
+    -- | Dispose of the 'Iterator' by closing any open handles.
+    --
+    -- Idempotent operation.
+  , iteratorClose :: HasCallStack => m (Either ImmutableDBError ())
+
+    -- | A identifier for the 'Iterator' that is unique for @m@.
+    --
+    -- This used for the 'Eq' instance, which is needed for testing.
+    --
+    -- TODO how can we avoid this abstraction leak?
```

Comment:
Can't we just use a pipe or other existing streaming abstraction?

Reply:
See `blobProducer` for an example of build a pipe (a `Producer`) using the iterator API.
The `Eq` instance is used for the quickcheck-state-machine tests, which are forthcoming.